crates/snops/src/state/external_peers.rs

```diff
@@ -0,0 +1,386 @@
+                    cache.update_peer_info(key, info.clone());
+                }
+            }
+        }
+
+        // wait 10 seconds between checks, including the time it took to process
+        tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    }
+}
+
+/// Get all online agents above a certain height in an environment
+pub fn online_agents_above_height(
+    state: &GlobalState,
+    env: EnvId,
+    height: u32,
+) -> Vec<AgentClient> {
+    let Some(env) = state.get_env(env) else {
+        return Vec::new();
+    };
+
+    env.node_peers
+        .iter()
+        .filter_map(|(_, peer)| {
+            // ensure peer is internal
+            let EnvPeer::Internal(agent_id) = peer else {
+                return None;
+            };
+            let agent = state.pool.get(agent_id)?;
+            // ensure peer height is above or equal the requested height
+            if agent.status.block_info.as_ref()?.height < height {
+                return None;
+            }
+            // ensure the agent is online
+            agent.client_owned()
+        })
+        .collect()
+}
+
+/// Obtain a peer's latest block hash and height
+/// We do not assume the hash and height are related, and they are used for
+/// separate purposes.
+async fn get_block_hash_for_peer(network: NetworkId, addr: SocketAddr) -> Option<(Arc<str>, u32)> {
+    // make a request to the external peer for the latest block hash
+    let hash_res = snarkos_request::get_on_addr::<Value>(network, "/block/hash/latest", addr)
+        .await
+        .ok()?;
+    let height_res = snarkos_request::get_on_addr::<Value>(network, "/block/hash/latest", addr)
+        .await
+        .ok()?;
+
+    Some((hash_res.as_str()?.into(), height_res.as_u64()? as u32))
+}
+
+const MAX_BLOCK_REQUEST_FAILURES: u8 = 3;
+
+/// Obtain a block from a list of agents, permits up to 3 failures
+async fn get_block_from_agents(
+    agents: &Vec<AgentClient>,
+    hash: ABlockHash,
+) -> Option<(LatestBlockInfo, Vec<ATransactionId>)> {
+    let mut failures = 0u8;
+    for agent in agents {
+        if let Ok(Some(block)) = agent.get_snarkos_block_lite(hash.to_string()).await {
+            return Some(block.split());
+        }
+        failures += 1;
+        if failures >= MAX_BLOCK_REQUEST_FAILURES {
+            break;
+        }
+    }
+    None
+}
+
+/// Obtain a peer's block info and transaction ids
+async fn get_block_info_for_peer(
+    network: NetworkId,
+    addr: SocketAddr,
+) -> Option<(LatestBlockInfo, Vec<ATransactionId>)> {
+    // make a request to the external peer for the latest block
+    // TODO: make this a RawValue to prevent unnecessarily parsing the response
+    let Ok(block_raw) = snarkos_request::get_on_addr::<Value>(network, "/block/latest", addr).await
+    else {
+        tracing::trace!("failed to get latest block for peer: {addr:?}");
+        return None;
+    };
+    let Some(block_hash) = block_raw.get("block_hash").and_then(|s| s.as_str()) else {
```

Comment:
You could turn all of these into `.or_else(|| { trace!("..."); None })?;`. It might take a `#[rustfmt::skip]` though and might not be worth it.

Reply:
going to leave as-is for now. might remove in a logging pass/refactor later